types/set.go

```diff
@@ -2,19 +2,94 @@ package types
 
 import (
 	"bytes"
+	"encoding/binary"
 	"encoding/json"
+	"hash/fnv"
+	"slices"
+
+	"golang.org/x/exp/maps"
 )
 
-// A Set is a collection of elements that can be of the same or different types.
-type Set []Value
+// A Set is an immutable collection of elements that can be of the same or different types.
+type Set struct {
+	s       map[uint64]Value
+	hashVal uint64
+}
+
+// NewSet takes a slice of Values and stores a clone of the values internally.
+func NewSet(v []Value) Set {
+	orig := make([]Value, len(v))
+	set := make(map[uint64]Value, len(v))
+
+	hashes := make([]uint64, 0, len(set))
+	for i, vv := range v {
+		orig[i] = vv
+
+		hash := vv.hash()
+		hashKey := hash
+
+		// Insert the value into the map. Deal with collisions via open addressing by simply incrementing the hash
+		// value. This method is safe so long as Set is immutable because nothing can be removed from the map.
+		for {
+			existing, ok := set[hashKey]
+			if !ok {
+				set[hashKey] = vv
+				hashes = append(hashes, hash)
+				break
+			} else if vv.Equal(existing) {
+				// found duplicate in slice
+				break
+			}
+			hashKey++
+		}
+	}
+
+	// Special case hashVal for empty set to 0 so that the return value of Value.hash() of Set{} and NewSet([]Value{})
+	// are the same
+	var hashVal uint64
+	if len(set) > 0 {
+		slices.Sort(hashes)
```

Comment:
Seems like we could probably just sum all the element's `hash()` and be faster without much additional fear of collisions? Basically, the hash() of other values becomes more important.

Note: I looked for reasons to believe this is true and found that Java implements `hashCode()` for `AbstractSet` this way:

> Returns the hash code value for this set. The hash code of a set is defined to be the sum of the hash codes of the elements in the set, where the hash code of a null element is defined to be zero. This ensures that s1.equals(s2) implies that s1.hashCode()==s2.hashCode() for any two sets s1 and s2, as required by the general contract of [Object.hashCode()](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#hashCode--).

Reply:
Good idea!